### ISSUE TITLE
Deadlock in multithread application

### ISSUE THREAD

**Issue by waihangcheung** [btraceio/btrace#539]
Our application is multithreaded, Sometimes we'd get a deadlock , may be related to #396 . We are using btrace v2.2.2

I have attached the stacktrace

```
Found one Java-level deadlock:
=============================
"Start Level: Equinox Container: d0df9587-30dc-001c-1612-83519f16ccd3":
  waiting to lock monitor 0x7fab3e5fe540 (object 0x3ff877e3d60, a org.eclipse.birt.core.framework.osgi.OSGILauncher$ChildFirstURLClassLoader),
  which is held by "Framework Event Dispatcher: Equinox Container: d0df9587-30dc-001c-1612-83519f16ccd3"
"Framework Event Dispatcher: Equinox Container: d0df9587-30dc-001c-1612-83519f16ccd3":
  waiting to lock monitor 0x7fab3e5fcf80 (object 0x203280bb3c0, a org.openjdk.btrace.instr.ClassCache),
  which is held by "Start Level: Equinox Container: d0df9587-30dc-001c-1612-83519f16ccd3"

Java stack information for the threads listed above:
===================================================
"Start Level: Equinox Container: d0df9587-30dc-001c-1612-83519f16ccd3":
	at org.eclipse.osgi.internal.loader.classpath.ClasspathEntry.getManifest(ClasspathEntry.java:110)
	at org.eclipse.osgi.internal.loader.classpath.ClasspathEntry.<init>(ClasspathEntry.java:59)
	at org.eclipse.osgi.internal.loader.ModuleClassLoader.createClassPathEntry(ModuleClassLoader.java:256)
	at org.eclipse.osgi.internal.loader.classpath.ClasspathManager.createClassPathEntry(ClasspathManager.java:331)
	at org.eclipse.osgi.internal.loader.classpath.ClasspathManager.getClasspath(ClasspathManager.java:283)
	at org.eclipse.osgi.internal.loader.classpath.ClasspathManager.addStandardClassPathEntry(ClasspathManager.java:200)
	at org.eclipse.osgi.internal.loader.classpath.ClasspathManager.addClassPathEntry(ClasspathManager.java:192)
	at org.eclipse.osgi.internal.loader.classpath.ClasspathManager.findClassPathEntry(ClasspathManager.java:175)
	at org.eclipse.osgi.internal.loader.classpath.ClasspathManager.buildClasspath(ClasspathManager.java:154)
	at org.eclipse.osgi.internal.loader.classpath.ClasspathManager.buildFragmentClasspaths(ClasspathManager.java:108)
	at org.eclipse.osgi.internal.loader.classpath.ClasspathManager.<init>(ClasspathManager.java:82)
	at org.eclipse.osgi.internal.loader.EquinoxClassLoader.<init>(EquinoxClassLoader.java:57)
	at org.eclipse.osgi.internal.loader.BundleLoader.createClassLoaderPrivledged(BundleLoader.java:277)
	at org.eclipse.osgi.internal.loader.BundleLoader.getModuleClassLoader(BundleLoader.java:224)
	- locked <0x0000020210576bd0> (a java.lang.Object)
	at org.eclipse.osgi.internal.loader.BundleLoader.findLocalResource(BundleLoader.java:795)
	at org.eclipse.osgi.internal.loader.sources.SingleSourcePackage.getResource(SingleSourcePackage.java:40)
	at org.eclipse.osgi.internal.loader.sources.MultiSourcePackage.getResource(MultiSourcePackage.java:42)
	at org.eclipse.osgi.internal.loader.BundleLoader.findResource(BundleLoader.java:575)
	at org.eclipse.osgi.internal.loader.ModuleClassLoader.getResource(ModuleClassLoader.java:201)
	at org.openjdk.btrace.instr.ClassInfo.inferClassLoader(ClassInfo.java:94)
	at org.openjdk.btrace.instr.ClassInfo.loadExternalClass(ClassInfo.java:202)
	at org.openjdk.btrace.instr.ClassInfo.<init>(ClassInfo.java:77)
	at org.openjdk.btrace.instr.ClassCache.get(ClassCache.java:66)
	- locked <0x00000203280bb3c0> (a org.openjdk.btrace.instr.ClassCache)
	at org.openjdk.btrace.instr.ClassInfo.loadExternalClass(ClassInfo.java:202)
	at org.openjdk.btrace.instr.ClassInfo.<init>(ClassInfo.java:77)
	at org.openjdk.btrace.instr.ClassCache.get(ClassCache.java:66)
	- locked <0x00000203280bb3c0> (a org.openjdk.btrace.instr.ClassCache)
	at org.openjdk.btrace.instr.ClassCache.get(ClassCache.java:58)
	at org.openjdk.btrace.instr.ClassFilter.isSubTypeOf(ClassFilter.java:165)
	at org.openjdk.btrace.instr.BTraceProbeSupport.getApplicableHandlers(BTraceProbeSupport.java:128)
	at org.openjdk.btrace.instr.BTraceProbePersisted.getApplicableHandlers(BTraceProbePersisted.java:409)
	at org.openjdk.btrace.instr.Instrumentor.create(Instrumentor.java:88)
	at org.openjdk.btrace.instr.BTraceClassWriter.addInstrumentor(BTraceClassWriter.java:73)
	- locked <0x00000202105872e8> (a java.util.ArrayDeque)
	at org.openjdk.btrace.instr.BTraceTransformer.transform(BTraceTransformer.java:156)
	at java.lang.instrument.ClassFileTransformer.transform(java.instrument@11.0.13.0.101/ClassFileTransformer.java:246)
	at sun.instrument.TransformerManager.transform(java.instrument@11.0.13.0.101/TransformerManager.java:188)
	at sun.instrument.InstrumentationImpl.transform(java.instrument@11.0.13.0.101/InstrumentationImpl.java:563)
	at java.lang.ClassLoader.defineClass1(java.base@11.0.13.0.101/Native Method)
	at java.lang.ClassLoader.defineClass(java.base@11.0.13.0.101/ClassLoader.java:1017)
	at org.eclipse.osgi.internal.loader.ModuleClassLoader.defineClass(ModuleClassLoader.java:272)
	at org.eclipse.osgi.internal.loader.classpath.ClasspathManager.defineClass(ClasspathManager.java:632)
	at org.eclipse.osgi.internal.loader.classpath.ClasspathManager.findClassImpl(ClasspathManager.java:588)
	at org.eclipse.osgi.internal.loader.classpath.ClasspathManager.findLocalClassImpl(ClasspathManager.java:540)
	at org.eclipse.osgi.internal.loader.classpath.ClasspathManager.findLocalClass(ClasspathManager.java:527)
	at org.eclipse.osgi.internal.loader.ModuleClassLoader.findLocalClass(ModuleClassLoader.java:324)
	at org.eclipse.osgi.internal.loader.BundleLoader.findLocalClass(BundleLoader.java:320)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClassInternal(BundleLoader.java:395)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:345)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:337)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.loadBundleActivator(BundleContextImpl.java:754)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.start(BundleContextImpl.java:706)
	at org.eclipse.osgi.internal.framework.EquinoxBundle.startWorker0(EquinoxBundle.java:936)
	at org.eclipse.osgi.internal.framework.EquinoxBundle$EquinoxModule.startWorker(EquinoxBundle.java:319)
	at org.eclipse.osgi.container.Module.doStart(Module.java:571)
	at org.eclipse.osgi.container.Module.start(Module.java:439)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1582)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1562)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:1533)
	- locked <0x000003ff33bf3310> (a java.lang.Object)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1476)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:340)
"Framework Event Dispatcher: Equinox Container: d0df9587-30dc-001c-1612-83519f16ccd3":
	at org.openjdk.btrace.instr.ClassCache.get(ClassCache.java:62)
	- waiting to lock <0x00000203280bb3c0> (a org.openjdk.btrace.instr.ClassCache)
	at org.openjdk.btrace.instr.ClassCache.get(ClassCache.java:58)
	at org.openjdk.btrace.instr.ClassFilter.isSubTypeOf(ClassFilter.java:165)
	at org.openjdk.btrace.instr.BTraceProbeSupport.getApplicableHandlers(BTraceProbeSupport.java:128)
	at org.openjdk.btrace.instr.BTraceProbePersisted.getApplicableHandlers(BTraceProbePersisted.java:409)
	at org.openjdk.btrace.instr.Instrumentor.create(Instrumentor.java:88)
	at org.openjdk.btrace.instr.BTraceClassWriter.addInstrumentor(BTraceClassWriter.java:73)
	- locked <0x000003ff53dbf5e8> (a java.util.ArrayDeque)
	at org.openjdk.btrace.instr.BTraceTransformer.transform(BTraceTransformer.java:156)
	at java.lang.instrument.ClassFileTransformer.transform(java.instrument@11.0.13.0.101/ClassFileTransformer.java:246)
	at sun.instrument.TransformerManager.transform(java.instrument@11.0.13.0.101/TransformerManager.java:188)
	at sun.instrument.InstrumentationImpl.transform(java.instrument@11.0.13.0.101/InstrumentationImpl.java:563)
	at java.lang.ClassLoader.defineClass1(java.base@11.0.13.0.101/Native Method)
	at java.lang.ClassLoader.defineClass(java.base@11.0.13.0.101/ClassLoader.java:1017)
	at java.security.SecureClassLoader.defineClass(java.base@11.0.13.0.101/SecureClassLoader.java:174)
	at org.eclipse.birt.core.framework.URLClassLoader.findClass1(URLClassLoader.java:181)
	at org.eclipse.birt.core.framework.URLClassLoader$1.run(URLClassLoader.java:156)
	at org.eclipse.birt.core.framework.URLClassLoader$1.run(URLClassLoader.java:1)
	at java.security.AccessController.doPrivileged(java.base@11.0.13.0.101/Native Method)
	at org.eclipse.birt.core.framework.URLClassLoader.findClass(URLClassLoader.java:151)
	at org.eclipse.birt.core.framework.osgi.OSGILauncher$ChildFirstURLClassLoader.loadClass(OSGILauncher.java:677)
	- locked <0x000002028930a950> (a java.lang.Class)
	at java.lang.ClassLoader.loadClass(java.base@11.0.13.0.101/ClassLoader.java:522)
	at org.eclipse.osgi.internal.log.EquinoxLogWriter.logged(EquinoxLogWriter.java:701)
	at org.eclipse.osgi.internal.log.ExtendedLogReaderServiceFactory.safeLogged(ExtendedLogReaderServiceFactory.java:86)
	at org.eclipse.osgi.internal.log.ExtendedLogReaderServiceFactory.logPrivileged(ExtendedLogReaderServiceFactory.java:205)
	at org.eclipse.osgi.internal.log.ExtendedLogReaderServiceFactory.log(ExtendedLogReaderServiceFactory.java:178)
	at org.eclipse.osgi.internal.log.LogServiceManager.frameworkEvent(LogServiceManager.java:106)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:926)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:340)

Found 1 deadlock.

```

[stack.log](https://github.com/btraceio/btrace/files/8772404/stack.log)

**Comment by waihangcheung** [comment]
May be relate to  

```
// ClassCache.java
  synchronized ClassInfo get(ClassLoader cl, ClassName className) {
    Map<ClassName, ClassInfo> infos = getInfos(cl);

    ClassInfo ci = infos.get(className);
    if (ci == null) {
      ci = new ClassInfo(this, cl, className);  // <-- this one may call get() again which would be blocked because of `synchronized`
      infos.put(className, ci);
    }
    return ci;
  }
```

```
// ClassInfo.java
  private void loadExternalClass(ClassLoader cl, ClassName className) {
  ...
        if (info.length > 1) {
            for (int i = 1; i < info.length; i++) {
              String ifc = info[i];
              if (ifc != null) {
                ClassName ifcClassName = new ClassName(ifc);
                supertypes.add(cache.get(inferClassLoader(cl, ifcClassName), ifcClassName));  <--- deadlock
              }
            }
          }
```

**Comment by jbachorik** [comment]
No, it's something different. The recursive call to `ClassCache.get()` would be done on the same thread so no deadlock can happen there.

The actual deadlock is because OSGI is using 2 different classloaders here
- 'org.eclipse.birt.core.framework.osgi.OSGILauncher$ChildFirstURLClassLoader`
- `org.eclipse.osgi.internal.loader.ModuleClassLoader`

The first classloader is used from the bootstrap and it actually locks on that instance.
The module class loader then attempts to load a class concurrently with the launcher classloader. Both of those attempts to load a class lead to the BTrace instrumenter being invoked and the first one to use the `ClassCache` then locks it. 

Unfortunately, the resolution mechanism asks for a resource from the classloader (which happens to be the `ModuleClassLoader` here) and that results in calling the `ChildFirstURLClassLoader` eventually.

At that moment the the `ChildFirstURLClassLoader` is already locked from the bootstrap so the resource resolution can not proceed. At the same time the bootstrap thread needs to access the `ClassCache` to do the instrumentation matching but it is locked by the `ModuleClassLoader`. And here we have the deadlock.

**Comment by jbachorik** [comment]
Could you apply #540 and build your BTrace binary to check if the fix resolves your problem?

I switched to concurrent maps and removed some unnecessary synchronization there.

**Comment by waihangcheung** [comment]
We tested several times, random hung in ConcurrentHashMap.tabAt (need kill -9 to terminate)

```
"main" #1 prio=5 os_prio=0 tid=0x620000000000 nid=0x13fb5  [ JVM thread_state=_thread_blocked, locked by VM at safepoint, polling bits: safep ]
   java.lang.Thread.State: RUNNABLE
	at java.util.concurrent.ConcurrentHashMap.tabAt(java.base@11.0.13.0.101/ConcurrentHashMap.java:760)
	at java.util.concurrent.ConcurrentHashMap.transfer(java.base@11.0.13.0.101/ConcurrentHashMap.java:2472)
	at java.util.concurrent.ConcurrentHashMap.addCount(java.base@11.0.13.0.101/ConcurrentHashMap.java:2348)
	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(java.base@11.0.13.0.101/ConcurrentHashMap.java:1773)
	at org.openjdk.btrace.instr.ClassCache.get(ClassCache.java:99)
	at org.openjdk.btrace.instr.ClassInfo.loadExternalClass(ClassInfo.java:202)
	at org.openjdk.btrace.instr.ClassInfo.<init>(ClassInfo.java:77)
	at org.openjdk.btrace.instr.ClassCache.lambda$get$0(ClassCache.java:99)
	at org.openjdk.btrace.instr.ClassCache$$Lambda$lambda$get$0$3092225380/0x000000000000095f.apply(Unknown Source)
	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(java.base@11.0.13.0.101/ConcurrentHashMap.java:1705)
	- locked <0x000010005e177088> (a java.util.concurrent.ConcurrentHashMap$ReservationNode)
	at org.openjdk.btrace.instr.ClassCache.get(ClassCache.java:99)
	at org.openjdk.btrace.instr.ClassCache.get(ClassCache.java:93)
	at org.openjdk.btrace.instr.ClassFilter.isSubTypeOf(ClassFilter.java:165)
	at org.openjdk.btrace.instr.BTraceProbeSupport.getApplicableHandlers(BTraceProbeSupport.java:128)
	at org.openjdk.btrace.instr.BTraceProbePersisted.getApplicableHandlers(BTraceProbePersisted.java:409)
	at org.openjdk.btrace.instr.Instrumentor.create(Instrumentor.java:88)
	at org.openjdk.btrace.instr.BTraceClassWriter.addInstrumentor(BTraceClassWriter.java:75)
	at org.openjdk.btrace.instr.BTraceTransformer.transform(BTraceTransformer.java:156)
	at java.lang.instrument.ClassFileTransformer.transform(java.instrument@11.0.13.0.101/ClassFileTransformer.java:246)
	at sun.instrument.TransformerManager.transform(java.instrument@11.0.13.0.101/TransformerManager.java:188)
	at sun.instrument.InstrumentationImpl.transform(java.instrument@11.0.13.0.101/InstrumentationImpl.java:563)
	at java.lang.ClassLoader.defineClass1(java.base@11.0.13.0.101/Native Method)
	at java.lang.ClassLoader.defineClass(java.base@11.0.13.0.101/ClassLoader.java:1017)
	at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.defineClass(BundleWiringImpl.java:2344)
	at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.defineClassParallel(BundleWiringImpl.java:2162)
	at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.findClass(BundleWiringImpl.java:2096)
	at org.apache.felix.framework.BundleWiringImpl.findClassOrResourceByDelegation(BundleWiringImpl.java:1565)
	at org.apache.felix.framework.BundleWiringImpl.access$300(BundleWiringImpl.java:79)
	at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.loadClass(BundleWiringImpl.java:1982)
	at java.lang.ClassLoader.loadClass(java.base@11.0.13.0.101/ClassLoader.java:522)
	at org.apache.felix.framework.Felix.loadBundleClass(Felix.java:2089)
	at org.apache.felix.framework.BundleImpl.loadClass(BundleImpl.java:978)
	at com.ullink.oms.osgi.common.ExtensionRegistrationProcessor.registerJibxClassesAndPreferences(ExtensionRegistrationProcessor.java:138)
	at com.ullink.oms.osgi.common.ExtensionRegistrationProcessor.registerJibxClassesAndPreferencesAndActivate(ExtensionRegistrationProcessor.java:740)
	- locked <0x0000100028790fc8> (a com.ullink.oms.osgi.common.ExtensionRegistrationProcessor)
	at com.ullink.oms.osgi.common.ExtensionsGovernor.registerBundle(ExtensionsGovernor.java:415)
	at com.ullink.oms.osgi.common.ExtensionsGovernor.preloadBundle(ExtensionsGovernor.java:590)
	at com.ullink.oms.osgi.common.ExtensionsGovernor.preloadBundles(ExtensionsGovernor.java:583)
	at com.ullink.oms.osgi.common.ExtensionsGovernor.preload(ExtensionsGovernor.java:104)
	at com.ullink.oms.Platform.configureAndStartServices(Platform.java:490)
	at com.ullink.oms.Platform.initialize(Platform.java:163)
	at com.ullink.oms.OrderManagementSystem.main(OrderManagementSystem.java:38)
```
[stack.log](https://github.com/btraceio/btrace/files/8797832/stack.log)

**Comment by jbachorik** [comment]
Thanks for testing!
Too bad it still does not work.

IMHO, this is a problem in OSGI - the `ChildFirstURLClassLoader` is locked, preventing concurrently obtaining resources from from that classloader. The module classloaders are operating with the assumption that the JVM is capable of parallel class definition. 
This setup makes it impossible to use any synchronization primitives in the class bytecode manipulation routines (the instrumentors).

I will try and see what can be done here but working this around will be an interesting problem ...

**Comment by jbachorik** [comment]
One quick attempt - I switched the `ConcurrentHashMap` which can call synchronized code sometimes to `NonBlockingHashMap` from the jctools project which is supposedly completely non-blocking.
I have updated the PR with that change so if you could try that out it would be awesome. I don't have any OSGI test app at hand and for non-osgi apps I can't reproduce the deadlock even with the original code.

**Comment by waihangcheung** [comment]
It fails to be compiled, can't found

```
import org.jctools.maps.NonBlockingHashMap;
import org.jctools.maps.NonBlockingIdentityHashMap;
```

**Comment by jbachorik** [comment]
Fixed in #540
